<a href="https://colab.research.google.com/github/JavokhirJambulov/AIapplicationSystem/blob/main/12204567_week6_hw.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Baseline Model:**

In [1]:
import tensorflow as tf

# Load Data
mnist = tf.keras.datasets.mnist

# Split data into train and test
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images / 255.0
test_images = test_images / 255.0

11490434/11490434 [==============================] - 0s 0us/step


**Activity 1:**

In [2]:
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.5),  # Add Dropout layer with a dropout rate
    tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5)

test_loss, test_acc = model.evaluate(test_images, test_labels)
model.save('mnist_cnn_model_with_dropout_bn.h5')

loaded_model = tf.keras.models.load_model('mnist_cnn_model_with_dropout_bn.h5')

test_loss, test_acc = loaded_model.evaluate(test_images, test_labels)

print("Test accuracy:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 99s 51ms/step - loss: 0.2085 - accuracy: 0.9397
Epoch 2/5
1875/1875 [==============================] - 81s 43ms/step - loss: 0.0931 - accuracy: 0.9736
Epoch 3/5
1875/1875 [==============================] - 82s 44ms/step - loss: 0.0751 - accuracy: 0.9793
Epoch 4/5
1875/1875 [==============================] - 88s 47ms/step - loss: 0.0607 - accuracy: 0.9828
Epoch 5/5
313/313 [==============================] - 5s 17ms/step - loss: 0.0351 - accuracy: 0.9894


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


313/313 [==============================] - 4s 12ms/step - loss: 0.0351 - accuracy: 0.9894
Test accuracy: 0.9894000291824341


 *Adding Dropout and Batch Normalization layers to the CNN model can significantly improve its performance. Dropout helps prevent overfitting by randomly dropping a fraction of neurons during training.*

 *Batch Normalization normalizes the activations, making the training more stable and accelerating convergence. These techniques are particularly useful when dealing with deep neural networks.*

**Test accuracy: 0.9894000291824341**

**Activity 2:**

In [5]:
# Define a list of activation functions to experiment with
activation_functions = ['sigmoid', 'tanh', 'relu']

for activation_function in activation_functions:
    # Build the CNN model with the specified activation function
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation=activation_function, input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Conv2D(64, (3, 3), activation=activation_function),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    # Compile the model
    model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])

    # Train the model
    model.fit(train_images, train_labels, epochs=5)

    # Evaluate the model
    test_loss, test_acc = model.evaluate(test_images, test_labels)

    print(f"Activation Function: {activation_function}")
    print("Test accuracy:", test_acc)
    print("------------------")

Epoch 1/5
1875/1875 [==============================] - 66s 35ms/step - loss: 0.6826 - accuracy: 0.7692
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.1219 - accuracy: 0.9634
Epoch 3/5
1875/1875 [==============================] - 61s 32ms/step - loss: 0.0781 - accuracy: 0.9762
Epoch 4/5
1875/1875 [==============================] - 61s 33ms/step - loss: 0.0594 - accuracy: 0.9814
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0562 - accuracy: 0.9817
Activation Function: sigmoid
Test accuracy: 0.9817000031471252
------------------
Epoch 1/5
1875/1875 [==============================] - 63s 33ms/step - loss: 0.1220 - accuracy: 0.9636
Epoch 2/5
1875/1875 [==============================] - 63s 34ms/step - loss: 0.0434 - accuracy: 0.9872
Epoch 3/5
1875/1875 [==============================] - 61s 33ms/step - loss: 0.0283 - accuracy: 0.9912
Epoch 4/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0229 - accuracy: 0

Activation Function: **sigmoid**

**Test accuracy: 0.9817000031471252**

------------------+------------------

Activation Function: **tanh**

**Test accuracy: 0.9847999811172485**

------------------+------------------

Activation Function: **relu**

**Test accuracy: 0.9908999800682068**

------------------+------------------

The choice of activation function can have a significant impact on the performance of a CNN. In the provided example, different activation functions (sigmoid, tanh, and ReLU) were experimented with. ReLU tends to perform well due to its ability to handle vanishing gradient problems, but the performance may vary depending on the dataset and network architecture. It's essential to experiment with activation functions to find the one that works best for a specific problem.

**Activity 3:**

In [4]:

model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(128, (3, 3), activation='relu'),  # Additional Conv2D layer
    tf.keras.layers.MaxPooling2D((2, 2)),  # Additional MaxPooling2D layer
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5)

test_loss, test_acc = model.evaluate(test_images, test_labels)
print("Test accuracy:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 60s 31ms/step - loss: 0.1749 - accuracy: 0.9460
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0624 - accuracy: 0.9807
Epoch 3/5
1875/1875 [==============================] - 61s 33ms/step - loss: 0.0446 - accuracy: 0.9856
Epoch 4/5
1875/1875 [==============================] - 63s 34ms/step - loss: 0.0339 - accuracy: 0.9894
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0556 - accuracy: 0.9829
Test accuracy: 0.9829000234603882


**Test accuracy: 0.9829000234603882**

Adding more convolutional and pooling layers can increase the model's capacity
to recognize complex features in the data. However, it also increases the model's computational complexity and may require more data to prevent overfitting. The impact on performance depends on the problem and the available data. Careful experimentation is necessary to determine the optimal depth and complexity of the CNN architecture.

**Activity 4:**

In [6]:
optimizers = ['sgd', 'rmsprop']

for optimizer in optimizers:
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5)

    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with {optimizer} optimizer:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 39s 20ms/step - loss: 0.4605 - accuracy: 0.8773
Epoch 2/5
1875/1875 [==============================] - 37s 20ms/step - loss: 0.2103 - accuracy: 0.9382
Epoch 3/5
1875/1875 [==============================] - 37s 20ms/step - loss: 0.1599 - accuracy: 0.9532
Epoch 4/5
1875/1875 [==============================] - 37s 20ms/step - loss: 0.1303 - accuracy: 0.9615
Epoch 5/5
313/313 [==============================] - 2s 7ms/step - loss: 0.1055 - accuracy: 0.9682
Test accuracy with sgd optimizer: 0.9682000279426575
Epoch 1/5
1875/1875 [==============================] - 41s 22ms/step - loss: 0.1491 - accuracy: 0.9554
Epoch 2/5
1875/1875 [==============================] - 41s 22ms/step - loss: 0.0526 - accuracy: 0.9839
Epoch 3/5
1875/1875 [==============================] - 41s 22ms/step - loss: 0.0345 - accuracy: 0.9893
Epoch 4/5
1875/1875 [==============================] - 42s 22ms/step - loss: 0.0248 - accuracy: 0.9926
Epoch 5/5
313/313 [=====

**Test accuracy with sgd optimizer: 0.9682000279426575**

**Test accuracy with rmsprop optimizer: 0.9879999756813049**


The choice of optimizer can affect the training speed and performance of a CNN. In the provided example, SGD and RMSprop optimizers were experimented with. SGD is a simple optimizer that may require fine-tuning of learning rates, while RMSprop adapts the learning rates and often converges faster. The choice of optimizer should consider the specific problem, and hyperparameter tuning may be necessary to achieve the best results.

**Activity 5:**

In [7]:
model = tf . keras . Sequential ([
    tf.keras.layers.Conv2D(32 , (3,3) , activation ='relu',input_shape =(28 , 28 , 1) ) ,
    tf.keras.layers.MaxPooling2D ((2 , 2) ) ,
    tf.keras.layers.Conv2D(64 , (3 ,3) , activation ='relu') ,
    tf.keras.layers.MaxPooling2D((2 , 2) ) ,
    tf.keras.layers.Flatten() ,
    tf.keras.layers.Dense(128 , activation ='relu') ,
    tf.keras.layers.Dense(10 , activation ='softmax')
])


model.compile( optimizer ='adam' ,loss = 'sparse_categorical_crossentropy',metrics = [ 'accuracy'])
model.fit( train_images , train_labels , epochs =5)

test_loss , test_acc = model . evaluate ( test_images , test_labels)
model.save ('mnist_cnn_model.h5')

loaded_model = tf.keras.models.load_model ('mnist_cnn_model.h5')

test_loss , test_acc = loaded_model.evaluate (test_images ,test_labels)

print("Test accuracy:", test_acc)

Epoch 1/5
1875/1875 [==============================] - 59s 31ms/step - loss: 0.1255 - accuracy: 0.9610
Epoch 2/5
1875/1875 [==============================] - 59s 32ms/step - loss: 0.0422 - accuracy: 0.9870
Epoch 3/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0285 - accuracy: 0.9909
Epoch 4/5
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0204 - accuracy: 0.9931
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 0.0300 - accuracy: 0.9912


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


313/313 [==============================] - 4s 11ms/step - loss: 0.0300 - accuracy: 0.9912
Test accuracy: 0.9911999702453613


**Test accuracy: 0.9911999702453613**

Evaluating the trained CNN model using test data is crucial to understand its performance on unseen data. The test accuracy provides a measure of how well the model generalizes to new samples. It helps in assessing the model's ability to make accurate predictions on real-world data and ensures that it has not overfit the training data. Evaluating the model with test data is a standard practice in machine learning to validate model performance.